conanfile.py

```diff
@@ -15,10 +15,10 @@ class VCMI(ConanFile):
         "minizip/[~1.2.12]",
     ]
     _clientRequires = [
-        "sdl/[~2.26.1 || >=2.0.20 <=2.22.0]", # versions in between have broken sound
-        "sdl_image/[~2.0.5]",
-        "sdl_mixer/[~2.0.4]",
-        "sdl_ttf/[~2.0.18]",
+        "sdl/[>=2.26.1 || >=2.0.20 <=2.22.0]", # versions in between have broken sound
+        "sdl_image/[>=2.0.5]",
+        "sdl_mixer/[>=2.0.4]",
+        "sdl_ttf/[>=2.0.18]",
```

Comment:
`>=` is dangerous because it allows to use completely untested v3. IIRC `^` locks major version,

Reply:
@kambala-decapitator 
I get what you're saying but not sure which approach would be better.

Examples of various 'bad' scenarios that we had over time:
- boost 1.86 - our conanfile permits 1.86, but it was not tested, so would have caused breakage if somebody were to use conan with no prebuilts package and conan downloaded 1.86
- sdl - for Android must be exact match, since its version is hardcoded in SDL wrapper for java code. So must be changed in sync. 
- sdl - 2.22 had broken sound bug we had some time ago, that first appeared on Debian
- sdl_image - there was hard to figure out regression on Flatpak due to broken png loading because of minor version update.
- ffmpeg seems to have rather stable API now, only one minor change despite 4.4 -> 7.0 jump, but was quite bad before.

So, approach that you're proposing is to set restrictions that are extremely unlikely to cause regressions.
Pros: less chance of regressions
Cons: if somebody were to try to build vcmi with newer libs he will run into bugs - 1.86 update on Arch, sdl_image update to 2.6 on Flatpak some time ago, sdl update to 2.22 on Debian before that. 

Alternatively, we could try to exclude versions that are known to cause problems and permit everything else. Combined with daily/weekly/monthly automated rebuild test job it should at the very least detect any new compilation / linking issues.
So in case of weird bug from some bleeding-edge system (Arch, Flatpak, unstable Debian) we could grab package from this scheduled rebuild and try to reproduce it.

Might change with artifactory / drop of prebuilts package, but at least right now it won't cause regressions as long as package *has* been tested - since we have `build=never` in CI, so conan will never grab newer version unless we added this newer version into the package.

Opinions?

Reply:
We can't be responsible for bugs in dependencies. Once we identify a buggy version, we can exclude it (like we already do with SDL) - I'm totally fine with that.

What I don't like is giving zero restrictions on future versions. Considering that most libs follow semver in one way or another, I'd rather lock major version of all dependencies, so that testing a new major version (e.g. SDL 3) is an explicit user's choice.

- Boost doesn't seem to follow semver at all as we see, but I don't think it's an often case that they break API like that
- SDL for android - well, if one updates only C++ part without the android one, then it's a problem of the user who didn't follow SDL manual

In the future would be useful to run Conan with newer libs versions to check if dependencies and project at least build, testing should also be done before updating binaries of course.

Reply:
Actually in case of SDL 3 it is not that we have not tested it and it may not work with vcmi.
We know *for sure* that it won't work with vcmi as it: https://github.com/libsdl-org/SDL/blob/main/docs/README-migration.md
Sure, we don't use 100% of SDL functions, but amount of breaking changes in this list is quite high.

So yeah, SDL 3 fits the definition of "exclude versions that don't work with vcmi", at least until we adapt code to SDL 3.

> SDL for android - well, if one updates only C++ part without the android one, then it's a problem of the user who didn't follow SDL manual

Kind of. Right now on develop if you simply generate conan package for Android from scratch (which our docs do allow) you will end up with different version of SDL in conan.

So enforcing correct version via conan makes sense. If you want to use different one - then update both conanfile and that java wrapper.

> Considering that most libs follow semver in one way or another

- sdl - major version breaks API. A lot.
- ffmpeg - no API breakage despite 4.4 -> 7.0 upgrade. Only one minor deprecation warning.
- boost - chance of API breakage in any version
- libtbb - uses year.month for releases, not semver
- qt - Not 100% sure, but I think at least with 5 -> 6 transition there was only one minor change that affected vcmi, so not really API breakage

So as far as our dependencies go, only SDL follow semver in terms of "change in major version number breaks API". Possibly also Qt, but we only use basic stuff from it.

Reply:
Qt 5->6 requires cmake changes at least for android. For others I agree.

Reply:
Changed. 
For SDL on PC we now request specifically 2.X. On mobiles - we request specific version.
Also added a comments to explain why.